src/t8_cad/t8_cad_shape_proximity.cxx

```diff
@@ -0,0 +1,307 @@
+          }
+          match = 1;
+          break;
+        }
+      }
+      if (!match) {
+        return 0;
+      }
+    }
+
+    if (!boundary)
+    {
+      /* Check if element centroid is inside shape (slow).
+       * This is still faster than checking if the element intersects the shape. 
+       * Only check this if boundary is false, because it would return true 
+       * for elements inside the shape. */
+      double              centroid[3] = { 0 };
+      t8_forest_element_centroid (forest, ltreeid, element, centroid);
+      if (t8_cad_shape_proximity::t8_cad_is_point_inside_shape (centroid, 0))
+      {
+        return 1;
+      }
+    }
+  }
+  /* TODO: Check if decomposition into solids and faces and bb checking for
+   * each individual brings a speedup */
+
+  /* Check for intersection of element and shape (very slow). */
+  gp_Pnt box_min(corner_coords[0], corner_coords[1], corner_coords[2]);
+  gp_Pnt box_max(corner_coords[3], corner_coords[4], corner_coords[5]);
+  TopoDS_Solid element_shape = BRepPrimAPI_MakeBox(box_min, box_max);
+  BRepExtrema_DistShapeShape dist_shape_shape;
+  dist_shape_shape.LoadS1(element_shape);
+  dist_shape_shape.LoadS2(occ_shape);
+  dist_shape_shape.Perform();
+  if(!dist_shape_shape.IsDone()){
+    SC_ABORTF("Failed to calculate distance between element and shape");
+  }
+  
+  /* Normally we would check if the distance is <= 0, but he have to use some tolerance,
+   * because otherwise OCC sorts out too many valid results. */
+  if (boundary){
+    if (dist_shape_shape.Value () <= Precision::Confusion())
+      return 2;
+    else if (dist_shape_shape.InnerSolution())
+      return 1;
+    else
+      return 0;
+  }
+  else {
+    return (dist_shape_shape.InnerSolution() ||
+            dist_shape_shape.Value () <= Precision::Confusion());
+  }
+}
+
+int
+t8_cad_shape_proximity::t8_cad_is_point_inside_shape (const double *coords, int optimize) const
+{
+  gp_Pnt pnt = gp_Pnt(coords[0], coords[1], coords[2]);
+
+  if (optimize)
+  {
+    /* Check if point is inside shape bounding box (very fast) */
+    if (occ_shape_bounding_box.IsOut(pnt)) {
+      return 0;
+    }
+
+    /* Check if point is inside subshape bounding box (fast) */
+    if (occ_shape_individual_bounding_boxes->Size() > 0) {
+      int match = 0;
+      for (auto it = occ_shape_individual_bounding_boxes->begin(); 
+           it != occ_shape_individual_bounding_boxes->end(); ++it)
+      {
+        if (!it->IsOut (pnt))
+        {
+          if (it != occ_shape_individual_bounding_boxes->begin())
+          {
+            const Bnd_OBB buffer_obb = *occ_shape_individual_bounding_boxes->begin();
+            occ_shape_individual_bounding_boxes->ChangeFirst() = *it;
+            *it = buffer_obb;
+          }
+          match = 1;
+          break;
+        }
+      }
+      if (!match) {
+        return 0;
+      }
+    }
+  }
+
+  /* Check if point is inside shape (slow) */
+  BRepClass3d_SolidClassifier classifier;
```

Comment:
If we do that, the function `t8_cad_is_point_inside_shape` cannot be `const` anymore, because the `Perform` function of the classifier is not `const`. What is more important, constness or the performance? Even though I do not think, that this has a significant increase in performance

Reply:
In my opinion we should optimize for performance here, as geometry evaluation can be rather expensive. Did you ever measure the cost of this function? (I know, this easier said than done). 
Maybe we should discuss this in the next t8code-Meeting.